### PR TITLE
Milvus 0.11 fpga support_2020_11_23_https://github.com/milvus-io/milvus/issues/3773

### PR DESCRIPTION
**Pull request by jackhouchina**
Signed-off-by: root <364427912@qq.com>

**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

Fixes #https://github.com/milvus-io/milvus/issues/3773

**What this PR does / why we need it:**

we want to support fpga device due to it performance

**Special notes for your reviewer:**

./build.sh -a -t Debug for fpga compile

**Additional documentation (e.g. design docs, usage docs, etc.):**

### CODE REVIEW

File: core/src/index/cmake/DefineOptionsCore.cmake [hunk 1]
```diff
@@ -10,7 +10,14 @@
...
 # or implied. See the License for the specific language governing permissions and limitations under the License.
 #-------------------------------------------------------------------------------
+#-------------------------------------------------
+set_option_category("FPGA version")
 
+if (MILVUS_FPGA_VERSION)
+    define_option(KNOWHERE_FPGA_VERSION "Build FPGA version" ON)
+else ()
+    define_option(KNOWHERE_FPGA_VERSION "Build FPGA version" OFF)
+endif ()
```
Comment (shengjun1985) [inline core/src/index/cmake/DefineOptionsCore.cmake:20]: It seems that "KNOWHERE_FPGA_VERSION" is never used.
May I remove it?

File: core/src/db/engine/ExecutionEngineImpl.h [hunk 2]
```diff
@@ -89,7 +92,7 @@ class ExecutionEngineImpl : public ExecutionEngine {
     ExecutionEngineContext context_;
 
     int64_t entity_count_;
-
+    std::mutex mutex_;
```
[the remark this answers is not in the document]
Comment (jackhouchina) [reply core/src/db/engine/ExecutionEngineImpl.h:95]: yes, please delete it .

File: core/src/db/engine/ExecutionEngineImpl.cpp [hunk 3]
```diff
@@ -147,6 +152,46 @@ ExecutionEngineImpl::Load(const TargetFields& field_names) {
...
+ExecutionEngineImpl::CopyToFpga() {
+#ifdef MILVUS_FPGA_VERSION
+    TimeRecorderAuto rc("ExecutionEngineImpl::CopyToFpga");
+    LOG_ENGINE_ERROR_ << "copy to fpga";
+    SegmentPtr segment_ptr;
+    segment_reader_->GetSegment(segment_ptr);
+
+    engine::VECTOR_INDEX_MAP new_map;
+    engine::VECTOR_INDEX_MAP& indice = segment_ptr->GetVectorIndice();
+    LOG_ENGINE_ERROR_ << indice.size() << "size";
```
[the remark this answers is not in the document]
Comment (jackhouchina) [reply core/src/db/engine/ExecutionEngineImpl.cpp:165]: please delete it .i forget to remove.

File: core/src/index/cmake/DefineOptionsCore.cmake (hunk 1 above)
Comment (jackhouchina) [reply core/src/index/cmake/DefineOptionsCore.cmake:20]: yes , you can.

**Review by shengjun1985 — APPROVED** [review]

**Pull request closed without merging**